File: gui/src/main/java/com/woohyman/gui/ui/preferences/ScreenViewPortSettingsActivity.java
package com.woohyman.gui.ui.preferences;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.KeyEvent;

import androidx.appcompat.app.AppCompatActivity;

import com.woohyman.gui.R;
import com.woohyman.gui.base.GameMenu;
import com.woohyman.gui.ui.multitouchbutton.MultitouchLayer;
import com.woohyman.keyboard.base.EmulatorUtils;
import com.woohyman.keyboard.base.SlotUtils;
import com.woohyman.keyboard.data.database.GameDescription;
import com.woohyman.keyboard.utils.DatabaseHelper;

import nostalgia.framework.data.entity.GfxProfile;
import nostalgia.framework.data.entity.SlotInfo;
import nostalgia.framework.utils.PreferenceUtil;

public class ScreenViewPortSettingsActivity extends AppCompatActivity
        implements GameMenu.OnGameMenuListener {

    MultitouchLayer mtLayer;
    String gameHash = "";
    DatabaseHelper dbHelper;
    Bitmap lastGameScreenshot;
    private GameMenu gameMenu;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.controler_layout);
        gameMenu = new GameMenu(this, this);
        mtLayer = findViewById(R.id.touch_layer);
        dbHelper = new DatabaseHelper(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mtLayer.setEditMode(MultitouchLayer.EDIT_MODE.SCREEN);
        GameDescription games = dbHelper.selectObjFromDb(GameDescription.class,
                "where lastGameTime!=0 ORDER BY lastGameTime DESC LIMIT 1");
        GfxProfile gfxProfile;
        lastGameScreenshot = null;

        if (games != null) {
            SlotInfo info = SlotUtils.getSlot(EmulatorUtils.getBaseDir(this),
                    games.checksum, 0);
            lastGameScreenshot = info.getScreenShot();
        }

        gfxProfile = PreferenceUtil.getLastGfxProfile(this);
        mtLayer.setLastgameScreenshot(lastGameScreenshot,
                gfxProfile == null ? null : gfxProfile.getName());
    }

    @Override
    protected void onPause() {
        super.onPause();
        mtLayer.saveScreenElement();
        mtLayer.stopEditMode();

        if (lastGameScreenshot != null) {
            lastGameScreenshot.recycle();
            lastGameScreenshot = null;
        }
    }


    @Override
    public void onGameMenuCreate(GameMenu menu) {
        menu.add(R.string.act_tcs_reset, R.drawable.ic_restart);
    }

    @Override
    public void onGameMenuPrepare(GameMenu menu) {
    }

    @Override
    public void onGameMenuOpened(GameMenu menu) {
    }

    @Override
    public void onGameMenuClosed(GameMenu menu) {
    }

    @Override
    public void onGameMenuItemSelected(GameMenu menu, GameMenuItem item) {
        runOnUiThread(() -> mtLayer.resetScreenElement());
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_MENU) {
            openGameMenu();
            return true;

        } else {
            return super.onKeyDown(keyCode, event);
        }
    }

    public void openGameMenu() {
        gameMenu.open();
    }

    @Override
    public void openOptionsMenu() {
        gameMenu.open();
    }

}

File: gui/xml/src/main/java/com/woohyman/xml/ui/preferences/ScreenViewPortSettingsActivity.kt
package com.woohyman.xml.ui.preferences

import android.graphics.Bitmap
import android.os.Bundle
import android.view.KeyEvent
import androidx.appcompat.app.AppCompatActivity
import com.woohyman.keyboard.base.EmulatorUtils
import com.woohyman.keyboard.base.SlotUtils
import com.woohyman.keyboard.data.database.GameDescription
import com.woohyman.keyboard.data.entity.GfxProfile
import com.woohyman.keyboard.utils.DatabaseHelper
import com.woohyman.keyboard.utils.PreferenceUtil.getLastGfxProfile
import com.woohyman.xml.R
import com.woohyman.xml.ui.menu.GameMenu
import com.woohyman.xml.ui.menu.GameMenu.GameMenuItem
import com.woohyman.xml.ui.menu.GameMenu.OnGameMenuListener
import com.woohyman.xml.ui.multitouchbutton.MultitouchLayer

class ScreenViewPortSettingsActivity : AppCompatActivity(), OnGameMenuListener {
    var mtLayer: MultitouchLayer? = null
    var gameHash = ""
    var dbHelper: DatabaseHelper? = null
    var lastGameScreenshot: Bitmap? = null

    private val gameMenu by lazy {
        GameMenu(this, this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.controler_layout)
        mtLayer = findViewById(R.id.touch_layer)
        dbHelper = DatabaseHelper(this)
    }

    override fun onResume() {
        super.onResume()
        mtLayer!!.editMode = MultitouchLayer.EDIT_MODE.SCREEN
        val games = dbHelper!!.selectObjFromDb(
            GameDescription::class.java,
            "where lastGameTime!=0 ORDER BY lastGameTime DESC LIMIT 1"
        )
        val gfxProfile: GfxProfile?
        lastGameScreenshot = null
        if (games != null) {
            val (_, _, _, screenShot) = SlotUtils.getSlot(
                EmulatorUtils.getBaseDir(this),
                games.checksum, 0
            )
            lastGameScreenshot = screenShot
        }
        gfxProfile = getLastGfxProfile(this)
        mtLayer!!.setLastgameScreenshot(
            lastGameScreenshot,
            gfxProfile.name!!
        )
    }

    override fun onPause() {
        super.onPause()
        mtLayer!!.saveScreenElement()
        mtLayer!!.stopEditMode()
        if (lastGameScreenshot != null) {
            lastGameScreenshot!!.recycle()
            lastGameScreenshot = null
        }
    }

    override fun onGameMenuCreate(menu: GameMenu) {
        menu.add(R.string.act_tcs_reset, R.drawable.ic_restart)
    }

    override fun onGameMenuPrepare(menu: GameMenu) {}
    override fun onGameMenuOpened(menu: GameMenu?) {

    }

    override fun onGameMenuClosed(menu: GameMenu) {}
    override fun onGameMenuItemSelected(menu: GameMenu?, item: GameMenuItem) {
        runOnUiThread { mtLayer!!.resetScreenElement() }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        return if (keyCode == KeyEvent.KEYCODE_MENU) {
            openGameMenu()
            true
        } else {
            super.onKeyDown(keyCode, event)
        }
    }

    fun openGameMenu() {
        gameMenu.open()
    }

    override fun openOptionsMenu() {
        gameMenu.open()
    }
}